### PR TITLE
Supported fee delegation login in caver.kct

### PR DESCRIPTION
**Pull request by sirano11**
## Proposed changes

This PR supported fee delegation feature in caver.kct package.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-java/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-java)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 1]
```diff
@@ -134,12 +134,45 @@ public static KIP17 deploy(Caver caver, String deployer, String name, String sym
     }
 
     /**
-     * Deploy a KIP-17 contract.
+     * Deploy a KIP-17 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
+     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in SendOptions.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:139]:
```suggestion
     * It must add deployer's keyring in `caver.wallet`.<p>
     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in `SendOptions` like below code example.
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 2]
```diff
@@ -134,12 +134,45 @@ public static KIP17 deploy(Caver caver, String deployer, String name, String sym
...
+     * <pre>
+     * <code>
+     *     SendOptions sendOptions = new SendOptions();
+     *     sendOptions.setFrom("deployer address");
+     *     sendOptions.setGas(BigInteger.valueOf(gas value));
+     *     sendOptions.setFeeDelegation(true);
+     *     sendOptions.setFeePayer("fee payer address");
+     *
+     *     KIP17 kip17 = KIP17.deploy(caver, sendOptions, name, symbol);
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:148]:
```suggestion
     *     KIP17 kip17 = caver.kct.kip17.deploy(caver, sendOptions, name, symbol);
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 3]
```diff
@@ -156,7 +189,41 @@ public static KIP17 deploy(Caver caver, String deployer, String name, String sym
     }
 
     /**
-     * Deploy KIP17 contract
+     * Deploy KIP-17 contract. <p>
+     * The wallet used in the contract is set to the wallet type passed as a parameter of the method.<p>
+     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in SendOptions.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:194]:
```suggestion
     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in `SendOptions` like below code example.
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 4]
```diff
@@ -156,7 +189,41 @@ public static KIP17 deploy(Caver caver, String deployer, String name, String sym
...
+     * <pre>
+     * <code>
+     *     SendOptions sendOptions = new SendOptions();
+     *     sendOptions.setFrom("deployer address");
+     *     sendOptions.setGas(BigInteger.valueOf(gas value));
+     *     sendOptions.setFeeDelegation(true);
+     *     sendOptions.setFeePayer("fee payer address");
+     *
+     *     KIP17 kip17 = KIP17.deploy(caver, sendOptions, name, symbol, caver.getWallet());
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:203]:
```suggestion
     *     KIP17 kip17 = caver.kct.kip17.deploy(caver, sendOptions, name, symbol, caver.getWallet());
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 5]
```diff
@@ -175,7 +242,39 @@ public static KIP17 deploy(Caver caver, KIP17DeployParams tokenInfo, String depl
     }
 
     /**
-     * Deploy KIP17 contract
+     * Deploy KIP-17 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
+     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in SendOptions.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:247]:
```suggestion
     * It must add deployer's keyring in `caver.wallet`.<p>
     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in `SendOptions` like below code example.
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 6]
```diff
@@ -175,7 +242,39 @@ public static KIP17 deploy(Caver caver, KIP17DeployParams tokenInfo, String depl
...
+     * <pre>
+     * <code>
+     *     SendOptions sendOptions = new SendOptions();
+     *     sendOptions.setFrom("deployer address");
+     *     sendOptions.setGas(BigInteger.valueOf(gas value));
+     *     sendOptions.setFeeDelegation(true);
+     *     sendOptions.setFeePayer("fee payer address");
+     *
+     *     KIP17DeployParams tokenInfo = new KIP17DeployParams(name, symbol);
+     *     KIP17 kip17 = KIP17.deploy(caver, tokenInfo, sendOptions);
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:257]:
```suggestion
     *     KIP17 kip17 = caver.kct.kip17.deploy(caver, tokenInfo, sendOptions);
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 7]
```diff
@@ -191,9 +290,45 @@ public static KIP17 deploy(Caver caver, KIP17DeployParams tokenInfo, String depl
...
+        sendOptions.setFrom(deployer);
+        sendOptions.setGas(BigInteger.valueOf(6500000));
+
+        return deploy(caver, tokenInfo, sendOptions, wallet);
+    }
+
+    /**
+     * Deploy KIP-17 contract.<p>
+     * The wallet used in the contract is set to the wallet type passed as a parameter of the method.<p>
+     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in SendOptions.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:303]:
```suggestion
     * If you want to deploy a contract using fee delegation transaction, you can create and send a fee delegated transaction through setting a fee delegation field in `SendOptions` like below code example.
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 8]
```diff
@@ -191,9 +290,45 @@ public static KIP17 deploy(Caver caver, KIP17DeployParams tokenInfo, String depl
...
+     * <pre>
+     * <code>
+     *     SendOptions sendOptions = new SendOptions();
+     *     sendOptions.setFrom("deployer address");
+     *     sendOptions.setGas(BigInteger.valueOf(gas value));
+     *     sendOptions.setFeeDelegation(true);
+     *     sendOptions.setFeePayer("fee payer address");
+     *
+     *     KIP17DeployParams tokenInfo = new KIP17DeployParams(name, symbol);
+     *     KIP17 kip17 = KIP17.deploy(caver, tokenInfo, sendOptions, caver.getWallet());
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:313]:
```suggestion
     *     KIP17 kip17 = caver.kct.kip17.deploy(caver, tokenInfo, sendOptions, caver.getWallet());
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 9]
```diff
@@ -1142,7 +1280,30 @@ private static SendOptions determineSendOptions(KIP17 kip17, SendOptions sendOpt
             value = sendOptions.getValue();
         }
 
-        newSendOptions = new SendOptions(from, gas, value);
+        if(feeDelegation || sendOptions.isFeeDelegation()) {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:1283]: `sendOptions ` that the user sends as parameters of the function should be reflected in higher priority than the default SendOptions(`getDefaultSendOptions()`).

Therefore, even if feeDelegation(from `getDefaultSendOptions()`) is `true`, it should be set `false` if `sendOptions.isFeeDelegation()` is `false`.

The default SendOptions should be the value used if the user has not set sendOptions.

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 10]
```diff
@@ -113,12 +110,12 @@ public KIP17(Caver caver, String contractAddress) throws IOException {
     }
 
     /**
-     * Deploy a KIP-17 contract.
+     * Deploy a KIP-17 contract.<p>
      * It must add deployer's keyring in caver.wallet.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:114]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`.
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 11]
```diff
@@ -134,12 +131,45 @@ public static KIP17 deploy(Caver caver, String deployer, String name, String sym
     }
 
     /**
-     * Deploy a KIP-17 contract.
+     * Deploy a KIP-17 contract.<p>
+     * It must add deployer's keyring in `caver.wallet`.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:135]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 12]
```diff
@@ -156,7 +186,41 @@ public static KIP17 deploy(Caver caver, String deployer, String name, String sym
...
+     * @throws TransactionException
+     */
+    public static KIP17 deploy(Caver caver, SendOptions sendOptions, String name, String symbol, IWallet wallet) throws NoSuchMethodException, IOException, InstantiationException, ClassNotFoundException, IllegalAccessException, InvocationTargetException, TransactionException {
+        KIP17DeployParams deployParams = new KIP17DeployParams(name, symbol);
+        return deploy(caver, deployParams, sendOptions, wallet);
+    }
+
+    /**
+     * Deploy KIP17 contract.<p>
      * It must add deployer's keyring in caver.wallet.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:224]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`.
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java [hunk 13]
```diff
@@ -175,7 +239,39 @@ public static KIP17 deploy(Caver caver, KIP17DeployParams tokenInfo, String depl
     }
 
     /**
-     * Deploy KIP17 contract
+     * Deploy KIP-17 contract.<p>
+     * It must add deployer's keyring in `caver.wallet`.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/KIP17.java:243]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/wrapper/KIP17Wrapper.java [hunk 14]
```diff
@@ -80,11 +82,94 @@ public KIP17 create(String contractAddress) throws IOException {
...
      */
     public KIP17 deploy(String deployer, String name, String symbol) throws NoSuchMethodException, IOException, InstantiationException, ClassNotFoundException, IllegalAccessException, InvocationTargetException, TransactionException {
-        return KIP17.deploy(this.caver, deployer, name, symbol, this.caver.getWallet());
+        return KIP17.deploy(this.caver, deployer, name, symbol);
     }
 
     /**
-     * Deploy KIP17 contract
+     * Deploy a KIP-17 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/wrapper/KIP17Wrapper.java:90]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip17/wrapper/KIP17Wrapper.java [hunk 15]
```diff
@@ -98,7 +183,89 @@ public KIP17 deploy(String deployer, String name, String symbol) throws NoSuchMe
      * @throws TransactionException
      */
     public KIP17 deploy(KIP17DeployParams tokenInfo, String deployer) throws NoSuchMethodException, IOException, InstantiationException, ClassNotFoundException, IllegalAccessException, InvocationTargetException, TransactionException {
-        return KIP17.deploy(this.caver, tokenInfo, deployer, this.caver.getWallet());
+        return KIP17.deploy(this.caver, tokenInfo, deployer);
+    }
+
+    /**
+     * Deploy KIP-17 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip17/wrapper/KIP17Wrapper.java:191]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip7/wrapper/KIP7Wrapper.java [hunk 16]
```diff
@@ -83,11 +85,103 @@ public KIP7 create(String contractAddress) throws IOException {
...
      */
     public KIP7 deploy(String deployer, String name, String symbol, int decimals, BigInteger initialSupply) throws NoSuchMethodException, IOException, InstantiationException, ClassNotFoundException, IllegalAccessException, InvocationTargetException, TransactionException {
-        return KIP7.deploy(this.caver, deployer, name, symbol, decimals, initialSupply, this.caver.getWallet());
+        return KIP7.deploy(this.caver, deployer, name, symbol, decimals, initialSupply);
     }
 
     /**
-     * Deploy KIP-7 contract.
+     * Deploy KIP-7 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip7/wrapper/KIP7Wrapper.java:93]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip7/wrapper/KIP7Wrapper.java [hunk 17]
```diff
@@ -101,7 +195,90 @@ public KIP7 deploy(String deployer, String name, String symbol, int decimals, Bi
      * @throws TransactionException
      */
     public KIP7 deploy(KIP7DeployParams tokenInfo, String deployer) throws NoSuchMethodException, TransactionException, IOException, InstantiationException, IllegalAccessException, InvocationTargetException, ClassNotFoundException {
-        return KIP7.deploy(this.caver, tokenInfo, deployer, this.caver.getWallet());
+        return KIP7.deploy(this.caver, tokenInfo, deployer);
+    }
+
+    /**
+     * Deploy KIP-7 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip7/wrapper/KIP7Wrapper.java:203]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java [hunk 18]
```diff
@@ -131,11 +128,43 @@ public static KIP37 deploy(Caver caver, String uri, String deployer) throws IOEx
...
+     * @return KIP37
+     * @throws IOException
+     * @throws NoSuchMethodException
+     * @throws InstantiationException
+     * @throws ClassNotFoundException
+     * @throws IllegalAccessException
+     * @throws InvocationTargetException
+     * @throws TransactionException
+     */
+    public static KIP37 deploy(Caver caver, String uri, SendOptions sendOptions) throws IOException, NoSuchMethodException, InstantiationException, ClassNotFoundException, IllegalAccessException, InvocationTargetException, TransactionException {
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java:157]: Is there any function that takes `KIP37DeployParams` as a parameter and deploys it?

Based on CommonArchitecture of the Caver, the deploy function of KCT(`caver.kct.kip7` or `caver.kct.kip17`) is to receive `KIPxxDeployParams` and deployer address as parameters.
Can KIP37 also have a function that can receive and deploy `KIP37DeployParams`?

File: core/src/main/java/com/klaytn/caver/kct/kip7/KIP7.java [hunk 19]
```diff
@@ -173,7 +244,39 @@ public static KIP7 deploy(Caver caver, KIP7DeployParams tokenInfo, String deploy
     }
 
     /**
-     * Deploy KIP-7 contract.
+     * Deploy KIP-7 contract.<p>
+     * It must add deployer's keyring in caver.wallet.<p>
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip7/KIP7.java:248]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`. <p>
```

File: core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java (hunk 18 above)
Comment (sirano11) [reply core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java:157]: The KIP-37 requires only one argument to deploy, I thought it need not to create a class KIP37DeployParams. 
Do you think the KIP37DeployParams needed?

File: core/src/main/java/com/klaytn/caver/kct/kip7/KIP7.java [hunk 20]
```diff
@@ -154,7 +189,43 @@ public static KIP7 deploy(Caver caver, String deployer, String name, String symb
...
+     * @throws IllegalAccessException
+     */
+    public static KIP7 deploy(Caver caver, SendOptions sendOptions, String name, String symbol, int decimals, BigInteger initialSupply, IWallet wallet) throws TransactionException, IOException, ClassNotFoundException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
+        KIP7DeployParams params = new KIP7DeployParams(name, symbol, decimals, initialSupply);
+        return deploy(caver, params, sendOptions, wallet);
+    }
+
+    /**
+     * Deploy KIP-7 contract.<p>
      * It must add deployer's keyring in caver.wallet.
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip7/KIP7.java:229]:
```suggestion
     * The deployer's keyring should be added in `caver.wallet`.
```

File: core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java (hunk 18 above)
Comment (jimni1222) [reply core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java:157]: This is to make the functional interface the same.
To deploy KCT, the user can think they need to pass the instance(`KIPxxDeployParams`) where the parameters can be defined and the address of the deployer (or `sendOptions`) as parameters according to the function interface of KCT deploy defined in common architecture of Caver.

Comment (sirano11) [reply core/src/main/java/com/klaytn/caver/kct/kip37/KIP37.java:157]: okay, I'll define it.

File: core/src/main/java/com/klaytn/caver/kct/kip37/KIP37DeployParams.java [hunk 21]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2020 The caver-java Authors
```
Comment (jimni1222) [inline core/src/main/java/com/klaytn/caver/kct/kip37/KIP37DeployParams.java:2]:
```suggestion
 * Copyright 2021 The caver-java Authors
```

**Review by jimni1222 — APPROVED** [review]
LGTM =) Thank you for changing!!!!

**Pull request merged**